libs/knowledge-store/ragstack_knowledge_store/links.py

```diff
@@ -0,0 +1,78 @@
+from dataclasses import dataclass
+from typing import Literal, Dict, Any, Set, Union
+import typing
+
+if typing.TYPE_CHECKING:
+    from langchain_core.documents import Document
+
+
+@dataclass(frozen=True)
+class Link:
+    kind: str
+    direction: Literal["incoming", "outgoing", "bidir"]
+
+    def __post_init__(self):
+        if self.__class__ == LinkTag:
+            raise TypeError("Abstract class Link cannot be instantiated")
+
+
+@dataclass(frozen=True)
+class LinkTag(Link):
+    tag: str
+
+
+@dataclass(frozen=True)
+class OutgoingLinkTag(LinkTag):
+    def __init__(self, kind: str, tag: str) -> None:
+        super().__init__(kind=kind, tag=tag, direction="outgoing")
+
+
+@dataclass(frozen=True)
+class IncomingLinkTag(LinkTag):
+    def __init__(self, kind: str, tag: str) -> None:
+        super().__init__(kind=kind, tag=tag, direction="incoming")
+
+
+@dataclass(frozen=True)
+class BidirLinkTag(LinkTag):
+    def __init__(self, kind: str, tag: str) -> None:
+        super().__init__(kind=kind, tag=tag, direction="bidir")
+
+
+LINKS = "links"
+
+
+def get_links(doc_or_md: Union["langchain_core.documents.Document", Dict[str, Any]]) -> Set[Link]:
+    """Get the links from a document or metadata.
+
+    Args:
+        doc_or_md: The metadata to get the link tags from.
+
+    Returns:
+        The set of link tags from the document or metadata.
+    """
+    try:
+        from langchain_core.documents import Document
```

Comment:
I don't see the advantage of supporting Document in get_links.
It's easy to do `get_links(doc.metadata)`

Reply:
It's not a dependency -- it's optional. If used from a project with Langchain it lets the user write `add_links(doc, x, y, z)` and more clearly conveys "add these links to the document" rather than `add_links(doc.metadata, x, y, z)`. I'm fine dropping this, but it seemed beneficial to let the user think about this as "adding links to the document".

Reply:
I feel like accepting documents here helps  communicate the API and makes for cleaner examples. Loose typing is pretty standard  in Python, for better or worse

Reply:
My feeling is that accepting 2 different types of input doesn’t make it clearer. You have to explain in comments that it’s « doc or metadata » and generally I would advocate for having 2 methods.
Anyway, not a hill I would die on. So it can be left that way.

Reply:
I'm going with this for the moment. I *think* I could be convinced the method should just move into `langchain` or the `langchain` package, at which point, the dependency on `Document` makes sense -- and maybe we drop support for metadata.

Reply:
Yes it should at least be moved to the langchain package.